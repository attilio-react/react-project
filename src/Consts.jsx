const consts = { 
    SEARCH_BY_TITLE : 0,
    SEARCH_BY_GENRE : 1,

    SORT_BY_RELEASE_DATE : 0,
    SORT_BY_RATING : 1
}

export {consts}
