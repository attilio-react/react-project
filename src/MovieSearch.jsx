import React from 'react';

import {Button} from './common/Button.jsx';
import {Label} from './common/Label.jsx';
import {InputText} from './common/InputText.jsx';
import {SearchByContainer} from './SearchByContainer.jsx';

class MovieSearch extends React.PureComponent {
  render () {
    return <>
		  <p><Label text='FIND YOUR MOVIE' /></p>
		  <p><InputText inputValue='input' /></p>
		  <p>
		     <SearchByContainer />
		     &nbsp;
		     &nbsp;
		     &nbsp;
		     &nbsp;
		     <Button caption='SEARCH' />
		  </p>
           </>;
  }
}


export {MovieSearch}


