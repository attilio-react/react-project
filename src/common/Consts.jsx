const consts = { 
    SEARCH_SCREEN : 0,
    DETAIL_SCREEN: 1,


    SEARCH_BY_TITLE : 0,
    SEARCH_BY_GENRE : 1,

    SORT_BY_RELEASE_DATE : 0,
    SORT_BY_RATING : 1,


    PARAM_SORT_BY: 'sortBy',
    PARAM_SORT_ORDER: 'sortOrder',
    PARAM_SEARCH: 'search',
    PARAM_SEARCH_BY: 'searchBy',

    VALUE_BY_GENRE: 'genres',
    VALUE_BY_TITLE: 'title',
    VALUE_BY_RELEASE_DATE: 'release_date',
    VALUE_BY_VOTE_AVERAGE: 'vote_average',

    VALUE_ASC: 'asc',
    VALUE_DESC: 'desc'
}

export {consts}
