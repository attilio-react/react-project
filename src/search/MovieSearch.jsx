import React from 'react';
import {connect} from 'react-redux'

import {consts} from 'Common/Consts.jsx'

import {SearchContext} from './SearchContext.jsx'

import {Button} from 'Common/Button.jsx';
import {Label} from 'Common/Label.jsx';
import {InputText} from 'Common/InputText.jsx';
import {SearchByContainer} from './SearchByContainer.jsx';
import {SearchResult} from './SearchResult.jsx';

import {
    searchBy, sortBy, searchTerm,
    getMoviesRequest} from 'Redux/ActionCreators.jsx'

class MovieSearchImpl extends React.PureComponent {
  render () {
    const {
            items, total,
            searchTerm, searchBy, sortBy,
            searchTermChangeCallback,
            searchButtonCallback, titleSearchButtonCallback, genreSearchButtonCallback,
            sortByReleaseDateCallback, sortByRatingCallback} = this.props,
	self = this


    return (<SearchContext.Provider value={{
            items: items,
            total: total,
            searchTerm: searchTerm,
            titleClickCb: () => titleSearchButtonCallback(self.props),
            genreClickCb: () => genreSearchButtonCallback(self.props),
            searchBySelection: searchBy,
            releaseDateClickCb: () => sortByReleaseDateCallback(self.props),
            ratingClickCb: () => sortByRatingCallback (self.props),
            sortBy: sortBy,
            itemClickCb: (id) => selectItemCallback(id)
	 }}>       
               <>
		  <p><Label text='FIND YOUR MOVIE' /></p>
		  <p><InputText onChange={searchTermChangeCallback} inputValue={searchTerm} /></p>
		  <p>
		     <SearchByContainer />
		     &nbsp;
		     &nbsp;
		     &nbsp;
		     &nbsp;
		     <Button onClick={() => searchButtonCallback(self.props)} caption='SEARCH' />
		  </p>
                  <SearchResult />
                </>
            </SearchContext.Provider>)
  }

}


const mapStateToProps = (state, ownProps) => {
      return {
                items: state.apiReducer.items,
                total: state.apiReducer.total,
                searchBy: state.guiReducer.searchBy,
                sortBy: state.guiReducer.sortBy,
                searchTerm: state.guiReducer.searchTerm
      }
}

const mapDispatchToProps = (dispatch, ownProps) => {
      const doGetMovies = (props, dispatch, override) => {
          const {searchTerm, searchBy, sortBy} = Object.assign({}, props, override)
          dispatch(getMoviesRequest({
              search: searchTerm,
              searchBy: (searchBy === consts.SEARCH_BY_GENRE ? consts.VALUE_BY_GENRE : consts.VALUE_BY_TITLE),
              sortBy: (sortBy === consts.SORT_BY_RELEASE_DATE ? consts.VALUE_BY_RELEASE_DATE : consts.VALUE_BY_VOTE_AVERAGE),
              sortOrder: consts.VALUE_ASC
          }))
      }
      return {
          searchButtonCallback: (props) => {
              doGetMovies(props, dispatch)
          },

          searchTermChangeCallback: (evt) => {
              dispatch(searchTerm(evt.target.value))
          },

          titleSearchButtonCallback: (props) => {
              dispatch(searchBy(consts.SEARCH_BY_TITLE))
              doGetMovies(props, dispatch, {searchBy: consts.SEARCH_BY_TITLE})
          },

          genreSearchButtonCallback: (props) => {
              dispatch(searchBy(consts.SEARCH_BY_GENRE))
              doGetMovies(props, dispatch, {searchBy: consts.SEARCH_BY_GENRE})
          },

          sortByReleaseDateCallback: (props) => {
              dispatch(sortBy(consts.SORT_BY_RELEASE_DATE))
              doGetMovies(props, dispatch, {sortBy: consts.SORT_BY_RELEASE_DATE})
          },

          sortByRatingCallback: (props) => {
              dispatch(sortBy(consts.SORT_BY_RATING))
              doGetMovies(props, dispatch, {sortBy: consts.SORT_BY_RATING})
          }
      }
}

const MovieSearch = connect(
      mapStateToProps,
      mapDispatchToProps
)(MovieSearchImpl)

export {MovieSearch}



