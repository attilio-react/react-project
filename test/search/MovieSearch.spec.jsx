import * as React from 'react';
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';

import {MovieSearch} from '|search/MovieSearch.jsx';
import {SearchContext} from '|search/SearchContext.jsx';
import {consts} from '|common/Consts.jsx'

const CONTEXT = {
    total: 50,
    searchTermCb: jest.fn(),
    searchClickCb: jest.fn(),
    searchTerm: 'it',
    items: []
}

describe('MovieSearch rendering', () => {
  it('Should render', () => {
    const 
        component = renderer.create(
            <SearchContext.Provider value={CONTEXT}>
               <MovieSearch />
            </SearchContext.Provider>),
        json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('MovieSearch behaviour', () => {
  it('Should call callback when search term changes', () => {
    const 
      resultHeader = mount(
            <SearchContext.Provider value={CONTEXT}>
               <MovieSearch />
            </SearchContext.Provider>),
      nativeInput = resultHeader.find('input').first();
    nativeInput.simulate('change', {target: {value: 'new value'}});
    expect(CONTEXT.searchTermCb).toBeCalled()
  });

  it('Should call callback when search button is clicked', () => {
    const 
      resultHeader = mount(
            <SearchContext.Provider value={CONTEXT}>
               <MovieSearch />
            </SearchContext.Provider>),
      nativeButton = resultHeader.find('input[value="SEARCH"]').first();
    nativeButton.simulate('click');
    expect(CONTEXT.searchClickCb).toBeCalled()
  });

});

