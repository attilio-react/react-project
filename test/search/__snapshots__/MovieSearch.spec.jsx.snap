// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieSearch rendering Should render 1`] = `
Array [
  <p>
    <span>
      FIND YOUR MOVIE
    </span>
  </p>,
  <p>
    <input
      onChange={[MockFunction]}
      type="text"
      value="it"
    />
  </p>,
  <p>
    <span>
      SEARCH BY
    </span>
     
    <input
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
      value="TITLE"
    />
     
    <input
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
      value="GENRE"
    />
           
    <input
      onClick={[MockFunction]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
      type="button"
      value="SEARCH"
    />
  </p>,
  <div
    style={
      Object {
        "fontSize": "large",
      }
    }
  >
    No films found
  </div>,
]
`;
